android/app/src/main/java/com/expensify/chat/CustomNotificationProvider.java

```diff
@@ -0,0 +1,234 @@
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+public class CustomNotificationProvider extends ReactNotificationProvider {
+
+    // Resize icons to 100 dp x 100 dp
+    private static final int MAX_ICON_SIZE_DPS = 100;
+
+    // Max wait time to resolve an icon. We have ~10 seconds to a little less
+    // to ensure the notification builds.
+    private static final int MAX_ICON_FETCH_WAIT_TIME_SECONDS = 8;
+
+    // Fallback drawable ID. 0 to not use a fallback ID.
+    private static final int FALLBACK_ICON_ID = 0;
+
+    // Logging
+    private static final String TAG = "NotificationProvider";
+
+    // Conversation JSON keys
+    private static final String PAYLOAD_KEY = "payload";
+    private static final String TYPE_KEY = "type";
+    private static final String REPORT_COMMENT_TYPE = "reportComment";
+
+    private final ExecutorService executorService = Executors.newCachedThreadPool();
+    public final HashMap<Integer, NotificationCache> cache = new HashMap<>();
+
+    public CustomNotificationProvider(@NonNull Context context, @NonNull AirshipConfigOptions configOptions) {
+        super(context, configOptions);
+    }
+
+    @NonNull
+    @Override
+    protected NotificationCompat.Builder onExtendBuilder(@NonNull Context context, @NonNull NotificationCompat.Builder builder, @NonNull NotificationArguments arguments) {
+        super.onExtendBuilder(context, builder, arguments);
+        PushMessage message = arguments.getMessage();
+
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+            return builder;
+        }
+
+        if (message.containsKey(PAYLOAD_KEY)) {
+            try {
+                JsonMap payload = JsonValue.parseString(message.getExtra(PAYLOAD_KEY)).optMap();
+
+                // Apply message style only for report comments
+                if (REPORT_COMMENT_TYPE.equals(payload.get(TYPE_KEY).getString())) {
+                        applyMessageStyle(builder, payload, arguments.getNotificationId());
+                }
+            } catch (Exception e) {
+                Log.e(TAG, "Failed to parse conversation", e);
+            }
+        }
+
+        return builder;
+    }
+
+    /**
+     * Applies the message style to the notification builder. It also takes advantage of the
+     * notification cache to build conversations.
+     *
+     * @param builder Notification builder that will receive the message style
+     * @param payload Notification payload, which contains all the data we need to build the notifications.
+     * @param notificationID Current notification ID
+     */
+    private void applyMessageStyle(NotificationCompat.Builder builder, JsonMap payload, int notificationID) {
+        int reportID = payload.get("reportID").getInt(-1);
```

Comment:
Shouldn't we return early here if the reportID is -1?

Reply:
Done!